contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +900,34 @@ clearly described and the permissions issues should be resolved.
 This could be done with a double opt-in (an opt-in from both the referrer and the refer-ee) or with secondary permissions
 checks performed in admission. 
 
-TODO: Plugins, extensions, nested kinds, headers
+TODO: Plugins, extensions, headers
 
+### Handling references to multiple kinds
+
+References which can refer to multiple kinds should use a single field, and select the target kind via `apiVersion` and `kind` fields.
+
+For example, if one can retrieving a referenced value from a `ConfigMap` or a `Secret` kind, the schema for the reference should be of the form:
+
+```yaml
+# preferred pattern
```

Comment:
If this is intended to be completely generic, how do you tell the reader of this reference what to do with the object it identifies? I.e., which field to read?

Reply:
the original PR intended to be a limited set of kinds, rather than completely generic.

But this would be a much better PR if there was an explanation of the various categories of object references, and the guidance of each. I'll be taking that as an action item.

Based on today's standards, the [recommedation is to provide a `fieldPath` field and use json-path like semantics to specify the field](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#objectreference-v1-core).

Reply:
Note that that doc doesn't mention json-path, you scared me for a minute there :) 

It's also not well-defined. It should probably explicitly choose JSONPointer, which can only reference particular fields, which I think is what we want anyway.

Reply:
The relevant point is:

> this string should contain a valid JSON/Go field access statement, such as desiredState.manifest.containers[2].

Which seems to imply, although not state, that jsonpath is a preferred mechanism (inferring from the path syntax). 

jsonPointer I think is valid (not strongly opinionated), and agreed it needs clarification. 

One point is that JSONPath is already used in [tools like kubectl](https://kubernetes.io/docs/reference/kubectl/jsonpath/), so it would provide some consistency to use JSONPath

Reply:
I would like to keep JSON Path out of the api wherever possible because it has no authoritative formal specification, and I think we want to eventually write a spec. For me this is a hard rule for the api machinery (e.g., never in a query parameter), and a nice-to-have for specific APIs.

JSONPointer is well specified, has basically the same syntax, and in this case its main limitation (referencing only a single field) is a feature as far as I can tell.

Reply:
> I would like to keep JSON Path out of the api wherever possible because it has no authoritative formal specification, and I think we want to eventually write a spec. For me this is a hard rule for the api machinery (e.g., never in a query parameter), and a nice-to-have for specific APIs.

Sounds good. I do see additional precedence for json-pointer as well, it looks like it's [what's passed into patch requests](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#other-resources).

It would be great if CLI and API could converge a single json specific syntax, but I do understand that jsonpath provides more a query-style syntax that could match multiple values.

Reply:
I've updated the PR to include examples of multiple scenarios (single resource reference, multi resource reference, field reference).

Also updated the guidance to use JSON Pointer for field path.